lib/charms/opensearch/v0/opensearch_performance_profile.py

```diff
@@ -0,0 +1,240 @@
+    """
+
+
+class OpenSearchPerformance(ops.Object):
+    """Base class for OpenSearch charms."""
+
+    _apply_profile_templates_event = EventSource(_ApplyProfileTemplatesOpenSearch)
+
+    def __init__(self, charm: ops.charm.CharmBase = None):
+        super().__init__(charm, None)
+        self.charm = charm
+        self.peers_data = self.charm.peers_data
+
+        self.framework.observe(
+            charm.on.install,
+            self._on_install,
+        )
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            charm.on[PeerRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            self._apply_profile_templates_event, self._on_apply_profile_templates
+        )
+        self._apply_profile_templates_has_been_called = False
+
+    def _on_install(self, event: EventBase):
+        """Handle install event.
+
+        Execute it on install, as defined in the docs, it is the first hook to run:
+           https://juju.is/docs/sdk/config-changed-event
+
+        install -> config-changed -> start
+
+        Store the current perf. profile we are applying. This must happen solely
+        on the beginning of the charm lifecycle. Therefore, we check if the unit is down.
+        """
+        if not os.path.exists(
+            os.path.join(
+                self.charm.opensearch.paths.conf,
+                self.charm.opensearch_config.JVM_OPTIONS,
+            )
+        ):
+            event.defer()
+            return
+        self.apply(self.charm.config.get(PERFORMANCE_PROFILE))
+
+    def _on_peer_cluster_relation_changed(self, _: EventBase):
+        """Handle the peer cluster relation changed event."""
+        if not (deployment_desc := self.charm.opensearch_peer_cm.deployment_desc()):
+            # Deployment description not available yet
+            # Nothing to do
+            return
+
+        if (
+            deployment_desc.typ != DeploymentType.MAIN_ORCHESTRATOR
+            and (rel_data := self.charm.opensearch_peer_cm.rel_data())
+            and rel_data.deployment_desc
+        ):
+            # Not the main orchestrator: set the profile from the peer-cluster
+            if self.apply(rel_data.deployment_desc.config.profile):
+                # We need to restart
+                self.charm.trigger_restart()
+            return
+
+        if self.apply(deployment_desc.config.profile):
+            # We need to restart
+            self.charm.trigger_restart()
+
+    @property
+    def current(self) -> OpenSearchPerfProfile | None:
+        """Return the current performance profile.
+
+        The profile is saved as a string in the charm peer databag.
+        """
+        if not self.peers_data.get(Scope.UNIT, PERFORMANCE_PROFILE) or not (
```

Comment:
1. why is this stored in the unit?
2. why is are we not just returning whatever's in the deployment_desc

Reply:
self.peers_data.get(Scope.UNIT, PERFORMANCE_PROFILE) -> what do I have configured right now? If I do not save it in the peer databag, then I have to figure it out using the meminfo + JVM heap + any template indices available. That is far more complex.

This is literally a "note for myself".

I cannot trust the `self.charm.config` either. As I may be right now changing my profile and the `self.charm.config` will represent the "new state" I have to be.

I cannot trust the deployment description: that is set by the leader unit. What happens if I am the last unit to be set and the deployment desc. has been updated?

Reply:
This setting is there to make it easier to find what is my real state, right now.